neo/rawio/medrawio.py

```diff
@@ -0,0 +1,358 @@
+            num_channels = len(self._stream_info[stream_index]['raw_chans'])
+            self.sess.set_reference_channel(self._stream_info[stream_index]['raw_chans'][0])
+        else:
+            if any(channel_indexes < 0):
+                raise IndexError(f'Can not index negative channels: {channel_indexes}')
+            # Set all channels to be inactive, then selectively set some of them to be active
+            self.sess.set_channel_inactive('all')
+            for i, channel_idx in enumerate(channel_indexes):
+                num_channels += 1
+                self.sess.set_channel_active(self._stream_info[stream_index]['raw_chans'][channel_idx])
+            self.sess.set_reference_channel(self._stream_info[stream_index]['raw_chans'][channel_indexes[0]])
+              
+        # Return empty dataset if start/stop sampes are equal
+        if i_start == i_stop:
+            raw_signals = np.zeros((0, num_channels), dtype='int32')
+            return raw_signals
+
+        # Otherwise, return the matrix 2D array returned by the MED library
+        start_sample_offset = self._stream_info[stream_index]['contigua'][seg_index]['start_index']
+        self.sess.read_by_index(i_start + start_sample_offset, i_stop + start_sample_offset)
+        
+        # Create "sample_major" 2D numpy array from the result of read_by_index()
+        samps_returned = len(self.sess.data['channels'][0]['data'])
+        raw_signals = np.array([], dtype=np.int32)
+        for chan in self.sess.data['channels']:
+            raw_signals = np.concatenate((raw_signals, chan['data']))
+        raw_signals = raw_signals.reshape(num_channels, samps_returned)
+        raw_signals = raw_signals.transpose()  # this makes is sample_major rather than channel_major
+        
+        return raw_signals
+
+    def _spike_count(self, block_index, seg_index, spike_channel_index):
+        return None
+
+    def _get_spike_timestamps(self, block_index, seg_index, spike_channel_index, t_start, t_stop):
+        return None
+
+    def _rescale_spike_timestamp(self, spike_timestamps, dtype):
+        return None
+
+    def _get_spike_raw_waveforms(self, block_index, seg_index, spike_channel_index,
+                                t_start, t_stop):
+        return None
+
+    def _event_count(self, block_index, seg_index, event_channel_index):
+    
+        # there are no epochs to consider for MED in this interface
+        if self.header['event_channels']['type'][event_channel_index] == b'epoch':
+            return 0
+                
+        records = self.sess.get_session_records()
+        
+        # Find segment boundaries
+        ts0 = (self.segment_t_start(block_index, seg_index) * 1e6) - self._session_time_offset
+        ts1 = (self.segment_t_stop(block_index, seg_index) * 1e6) - self._session_time_offset
+        
+        # Only count Note and Neuralynx type records.
+        count = 0
+        for record in records:
```

Comment:
I agree that this may not be the most efficient way to do this, but in MED there are potentially an infinite number of types of events, although only a handful have been defined so far.  In Neo, all events have to be represented as strings, whereas in MED, they can be arbitrary binary structures.   Each event type supported by Neo would need a "to_string" function.
One of the obstacles I ran into with the Neo wrapper, in general, is the distinction between retrieving samples (analogsignal_chunk) and retrieving events.  The MED way of doing this, in the dhn_med_py library, it to retrieve both at the same time with a single call.  This matches the use-case for a data viewer, where you might want 30 seconds of data and the same 30 seconds of events.
For the purposes of the Neo wrapper, we created a separate get_session_records() function within dhn_med_py, which only retrives records for a certain time range.  That's the call being used here.

Reply:
For the counting of events in a stream/segment, I could move that loop to the parse header and cache the results (cache the count number).  If you think all the data of the events should be cached, we can do that too - although that is potentially a lot of data, as there could be automated events (spikes, seizure detections, or hardware generated events).  The dhn_med_py library's get_session_records(t1, t2) function is actually pretty fast, if a narrow time slice is specified, since only looks in the relevant sections of data to retrieve the events.  Internally, all MED data is divided into segments (which is a slightly different concept than Neo's segments).  When reading samples or events from MED, only the relevant segments are searched, and in the case of samples, caching is done for already decoded data within the dhn_med_py library to speed it up.

There might be some tricks we could do for counting events (without actually looking at the data in them).  That's not currently a part of dhn_med_py, as we didn't anticipate user need for that.

Reply:
Ok, let's keep the idea of counting the number of events already in the `parse_header` it a need for performance optimisation arises in the future.